File: mpi_example4.c
#include <stdio.h>
#include <mpi.h>

int main(int argc, char *argv[]) {
  
  int numtasks, my_rank, prev, next;
  MPI_Status status;
  
  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &numtasks);
  MPI_Comm_rank(MPI_COMM_WORLD, &my_rank);
  
  prev = my_rank - 1;
  next = my_rank + 1;
  
  if (my_rank == 0) {
    prev = numtasks - 1;
  }
  if (my_rank == numtasks - 1) {
    next = 0;
  }
  
  int sbuf = my_rank;
  int rbuf1, rbuf2;
  
  // Blocking sends
  MPI_Send(&sbuf, 1, MPI_INT, prev, 2, MPI_COMM_WORLD);
  MPI_Send(&sbuf, 1, MPI_INT, next, 1, MPI_COMM_WORLD);
  
  // Blocking receives
  MPI_Recv(&rbuf1, 1, MPI_INT, prev, 1, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
  MPI_Recv(&rbuf2, 1, MPI_INT, next, 2, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
    
  
  printf("My rank: %d and my neighbors: %d %d\n", my_rank, rbuf1, rbuf2);
  
  MPI_Finalize();
  return 0;
}

File: mpi_example4.py
from mpi4py import MPI
import numpy as np

def main():
    
    comm = MPI.COMM_WORLD
    my_rank = comm.Get_rank()
    nprocs = comm.Get_size()

    prev = my_rank - 1
    next = my_rank + 1

    if my_rank == 0:
        prev = nprocs - 1
    if my_rank == nprocs - 1:
        next = 0

    rbuf1 = np.empty(1, dtype=int)
    rbuf2 = np.empty(1, dtype=int)
    sbuf = np.array([my_rank], dtype=int)
    
    tag1 = 1
    tag2 = 2

    # blocking sends
    comm.Send([sbuf, MPI.INT], dest=prev, tag=tag2)
    comm.Send([sbuf, MPI.INT], dest=next, tag=tag1)
    # blocking receives
    comm.Recv([rbuf1, MPI.INT], source=prev, tag=tag1)
    comm.Recv([rbuf2, MPI.INT], source=next, tag=tag2)

    print("My rank: {:d} and my neighbors: {:d} {:d}".format(my_rank, rbuf1[0], rbuf2[0]))
    
if __name__ == "__main__":
    main()

